File: 0001-0500/0101-Symmetric-Tree/py-0101/Solution1.py
# Source : https://leetcode.com/problems/symmetric-tree/
# Author : penpenps
# Time   : 2019-07-09

# Revert right node firstly, then compare with left node
# Time Complexity: O(n)
# Space Complexity: O(n)

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        # Revert TreeNode
        def revert(node):
            if not node: return
            node.left, node.right = node.right, node.left
            revert(node.left)
            revert(node.right)
        
        def isEqual(left, right):
            if not left and not right: return True
            if not left or not right: return False
            if left.val != right.val:
                return False
            return isEqual(left.left, right.left) and isEqual(left.right, right.right)
        
        if not root:
            return True
        revert(root.right)
        return isEqual(root.left, root.right)

File: 0001-0500/0226-Invert-Binary-Tree/java-0226/src/Solution1.java
/// Source : https://leetcode.com/problems/invert-binary-tree/description/
/// Author : liuyubobobo
/// Time   : 2017-11-17

/// Recursive
/// Time Complexity: O(n), where n is the node's number of the tree
/// Space Complexity: O(h), where h is the height of the tree
public class Solution1 {

    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    public TreeNode invertTree(TreeNode root) {

        if(root == null)
            return null;

        TreeNode left = invertTree(root.left);
        TreeNode right = invertTree(root.right);

        root.left = right;
        root.right = left;

        return root;
    }
}
